Fix: Custom Dimension not synced and created after enabling Audience …

ankitrox opened the pull request:
…Segmentation without dashboard refresh.

## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #9597 

## Relevant technical choices

<!-- Please describe your changes. -->

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [ ] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.
- [x] Ensure there are no unexpected significant changes to file sizes.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

nfmohit reviewed (changes requested):
Thank you for your work here, @ankitrox. Please take a look at the following points:

1. I've left some CR feedback. The first one is quite major as the change does not align with the IB.
2. Please add a QAB, thanks!

Please let me know if you have any questions, thank you.

techanvil commented on assets/js/modules/analytics-4/datastore/audiences.js:
```diff
@@ -383,9 +383,7 @@ const baseActions = {
-			yield commonActions.await(
-				dispatch( MODULES_ANALYTICS_4 ).syncAvailableAudiences()
-			);
+			yield fetchSyncAvailableAudiencesStore.actions.fetchSyncAvailableAudiences();
```
Thanks @nfmohit, you are right - we should _add_ a call to yield `fetchSyncAvailableCustomDimensionsStore.actions.fetchSyncAvailableCustomDimensions()` and return its error if it returns one, rather than replace the call to `syncAvailableAudiences()`.

nfmohit replied:
Thank you @techanvil !

nfmohit replied:
@ankitrox, I believe we should revert this change and dispatch the `syncAvailableAudiences` action as before. This change is not necessary.

Additionally, the call to `fetchSyncAvailableCustomDimensions` added in the `syncAvailableAudiences` action should be moved here, right before/after the dispatch to `syncAvailableAudiences`.

nfmohit commented on assets/js/modules/analytics-4/datastore/audiences.js:
```diff
@@ -198,6 +198,17 @@ const baseActions = {
 			return { response: availableAudiences, error };
 		}
 
+		const {
+			response: availableCustomDimensions,
+			error: syncDimensionsError,
+		} = yield commonActions.await(
+			dispatch( MODULES_ANALYTICS_4 ).fetchSyncAvailableCustomDimensions()
+		);
+
+		if ( syncDimensionsError ) {
+			return { response: availableCustomDimensions, error };
+		}
+
```
Could you help me understand this change? The IB mentions the available custom dimensions should be synced within the `enableAudienceGroupMain` action. I don't think we need to sync custom dimensions whenever available audiences are synced.

ankitrox commented:
@nfmohit Extremely sorry, the changes were added in the wrong place, but those were working during testing so did not get noticed.

I have changed this and made changes in `enableAudienceGroupMain` action as IB mentions. Also, adjusted the failing tests as well as we already have the test for ensuring that custom dimension gets created when it does not exist here: https://github.com/google/site-kit-wp/blob/4d66d5b27f10bd0555c50b0f4b75cc73bd9e0acc/assets/js/modules/analytics-4/datastore/audiences.test.js#L1388-L1491

* I have tested the functionality which is working fine.
* Added the QAB and test is working as expected as per QAB steps.
* All JS tests are passing and VRT tests are passing locally.

Over to you for another look. Sorry again for inconvenience @nfmohit 😞

nfmohit reviewed (approved):
LGTM, nice work, @ankitrox !
